[tp] add kwargs support to prepare_module_input

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124114

as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

wanchaol committed:
[tp] add kwargs support to prepare_module_input

as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

[ghstack-poisoned]

pbelevich commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -405,8 +416,32 @@ def _prepare_input_fn(self, inputs, device_mesh):
+                if desired_layout is not None and desired_layout != input_layout:
+                    kwarg_val = kwarg_val.redistribute(placements=(desired_layout,))
+
+            prepared_kwarg_inputs[kwarg_key] = kwarg_val.to_local() if self.use_local_output else kwarg_val
```
`kwarg_val` can be anything(`None` or whatever type here) and it fails on HF models

wanchaol replied:
should be fixed now

wanchaol committed:
Update on "[tp] add kwargs support to prepare_module_input"


as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol committed:
Update on "[tp] add kwargs support to prepare_module_input"


as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol committed:
Update on "[tp] add kwargs support to prepare_module_input"


as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -372,24 +378,37 @@ class PrepareModuleInput(ParallelStyle):
         *,
-        input_layouts: Union[Placement, Tuple[Placement]],
-        desired_input_layouts: Union[Placement, Tuple[Placement]],
+        input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
+        desired_input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
```
curious: does ParallelStyle assume 1-d mesh? It seems so.

wanchaol replied:
yeah the TP API currently assumes 1-d mesh

XilunWu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -372,24 +378,37 @@ class PrepareModuleInput(ParallelStyle):
-        desired_input_layouts: Union[Placement, Tuple[Placement]],
+        input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
+        desired_input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
+        input_kwarg_layouts: Optional[Dict[str, Placement]] = None,
+        desired_input_kwarg_layouts: Optional[Dict[str, Placement]] = None,
```
Can kwarg layouts include non-Tensor arguments?

XilunWu reviewed (approved):
lgtm. Added 2 questions. Besides, a similar logic should apply to `local_map` as well? So far `local_map` does not process `kwargs`.

wanchaol commented:
> lgtm. Added 2 questions. Besides, a similar logic should apply to `local_map` as well? So far `local_map` does not process `kwargs`.

🤔 yeah I think we can interate the `local_map` API to include kwargs, does not need to be in the initial PR but it's a good thing to think about how to solve kwargs

wanchaol commented:
@pytorchbot merge
